multiversx_sdk/network_providers/transaction_decoder.py

```diff
@@ -0,0 +1,236 @@
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "sender": self.sender,
+            "receiver": self.receiver,
+            "value": self.value,
+            "function_name": self.function_name if self.function_name else "",
+            "function_args": self.function_args if self.function_args else [],
+            "transfers": self._transfers_to_dict()
+        }
+
+    def _transfers_to_dict(self) -> List[Dict[str, Any]]:
+        if self.transfers:
+            if not len(self.transfers):
+                return []
+
+            transfers: List[Dict[str, Any]] = []
+
+            for transfer in self.transfers:
+                transfers.append({
+                    "value": transfer.amount,
+                    "token": transfer.token.identifier,
+                    "nonce": transfer.token.nonce
+                })
+
+            return transfers
+
+        return []
+
+
+class TransactionDecoder:
+    def get_transaction_metadata(self, transaction: ITransactionToDecode) -> TransactionMetadata:
+        metadata = self.get_normal_transaction_metadata(transaction)
+
+        esdt_metadata = self.get_esdt_transaction_metadata(metadata)
+        if esdt_metadata:
+            return esdt_metadata
+
+        nft_metadata = self.get_nft_transfer_metadata(metadata)
+        if nft_metadata:
+            return nft_metadata
+
+        multi_metadata = self.get_multi_transfer_metadata(metadata)
+        if multi_metadata:
+            return multi_metadata
+
+        return metadata
+
+    def get_normal_transaction_metadata(self, transaction: ITransactionToDecode) -> TransactionMetadata:
+        metadata = TransactionMetadata()
+        metadata.sender = transaction.sender.to_bech32()
+        metadata.receiver = transaction.receiver.to_bech32()
+        metadata.value = transaction.value
+
+        if transaction.data:
+            data_components = transaction.data.split("@")
+
+            args = data_components[1:]
+            if all(self.is_smart_contract_call_argument(x) for x in args):
+                metadata.function_name = data_components[0]
+                metadata.function_args = args
+
+        return metadata
+
+    def get_esdt_transaction_metadata(self, metadata: TransactionMetadata) -> Optional[TransactionMetadata]:
+        if metadata.function_name != "ESDTTransfer":
+            return None
+
+        args = metadata.function_args
+        if not args:
+            return None
+
+        if len(args) < 2:
+            return None
+
+        identifier = self.hex_to_string(args[0])
+        value = self.hex_to_number(args[1])
+
+        result = TransactionMetadata()
+
+        result.sender = metadata.sender
+        result.receiver = metadata.receiver
+        result.value = value
+        result.transfers = []
+
+        if len(args) > 2:
+            result.function_name = self.hex_to_string(args[2])
+            result.function_args = args[3:]
```

Comment:
the `if` check does not cover args[3:] if len if 3. will it throw an error, or simply return an empty array?

Reply:
it returns an empty array